[Move documentation] Update book to use abilities (#8582)
diem/diem: language/documentation/book/src/structs-and-resources.md

Structs cannot be recursive, so the following definition is invalid:

```move=
struct Foo { x: Foo }
//              ^ error! Foo cannot contain Foo
```

> As mentioned above: by default, a struct declaration is linear and ephemeral. So to allow the value to be used with certain operations (that copy it, drop it, store it in global storage, or use it as a storage schema). To grant [abilities](./abilities.md), structs can be annotated with `has <ability>`

Reviewer: ```suggestion
As mentioned above: by default, a struct declaration is linear and ephemeral. So, to allow the value to be used with certain operations (that copy it, drop it, store it in global storage, or use it as a storage schema), structs can be granted [abilities](./abilities.md) by annotating them with `has <ability>`:
```

```move=
address 0x2 {
module M {
    struct Foo has copy, drop { x: u64, y: bool }
}
}
```

> For more details, see the section in abilities on [annotating structs](./abilities.md#annotating-structs)

Reviewer: ```suggestion
For more details, see the [annotating structs](./abilities.md#annotating-structs) section.
```

If you need to read and copy a field's value, you can then dereference the borrowed field

```move=
let foo = Foo { x: 3, y: true };
let bar = Bar { foo: copy foo };
let x: u64 = *&foo.x;
let y: bool = *&foo.y;
let foo2: Foo = *&bar.foo;
```

> If the field is an implicitly copyable, the dot operator can be used to read fields of a struct without any borrowing. (Only scalar values with the `copy` ability are implicitly copyable)

Reviewer: ```suggestion
If the field is implicitly copyable, the dot operator can be used to read fields of a struct without any borrowing. (Only scalar values with the `copy` ability are implicitly copyable.)
```

## Ownership

As mentioned above in [Defining Structs](#defining-structs), structs are by default linear and ephemeral. This means they cannot be copied or dropped. This property can be very useful when modeling real world resources like money, as you do not want money to be duplicated or get lost in circulation.

```move=
address 0x2 {
module M {
    struct Foo { x: u64 }

    public fun copying_resource() {
        let foo = Foo { x: 100 };
        let foo_copy = copy foo; // error! cannot be copy without the 'copy' ability
        let foo_ref = &foo;

>         let another_copy = *foo_ref /// error! cannot be copy without the 'copy' ability

Reviewer: Why 3 slashes before the comment? This comment doesn't seem quite right, either. It seems awkward to say "cannot be copy" (also above), and in this case, the copy is only implied by the dereference. The actual error message is: "Invalid dereference. Dereference requires the 'copy' ability". The last part of that message would fit well here.

Author: Cleaned up the surrounding ones too

    }

    public fun destroying_resource1() {
        let foo = Foo { x: 100 };

        // error! when the function returns, foo still contains a value without the 'drop' ability
    }

    public fun destroying_resource2(f: &mut Foo) {
        *f = Foo { x: 100 } // error! destroying a value without the 'drop' ability
    }
}
}
```

To fix the second example (`fun dropping_resource`), you would need to manually "unpack" the resource:

```move=
address 0x2 {
module M {
    struct Foo { x: u64 }

    public fun destroying_resource1_fixed() {
        let foo = Foo { x: 100 };
        let Foo { x: _ } = foo;
    }
}
}
```

Recall that you are only able to deconstruct a resource within the module in which it is defined.
This can be leveraged to enforce certain invariants in a system, for example, conservation of money.

If on the other hand, your struct does not represent something valuable, you can add the abilities
`copy` and `drop` to get a struct value that might feel more familiar from other programming

> languages

Reviewer: ```suggestion
languages.
```

```move=
address 0x2 {
module M {
    struct Foo has copy, drop { x: u64 }

    public fun run() {
        let foo = Foo { x: 100 };
        let foo_copy = copy foo;
        // ^ this code copies foo, whereas `let x = foo` or
        // `let x = move foo` both move foo

        let x = foo.x;            // x = 100
        let x_copy = foo_copy.x;  // x = 100

        // both foo and foo_copy are implicitly discarded when the function returns
    }
}
}
```

## Storing Resources in Global Storage

Only structs with the `key` ability can be saved directly in [persistent global storage](./global-storage-operators.md). All values stored within those `key` structs must have the `store` abilities. See the [ability](./abilities] and [global storage](./global-storage-operators.md) chapters for more detail.

## Examples

Here are two short examples of how you might use structs to represent valuable data (in the case of `Coin`) or more classical data (in the case of `Point` and `Circle`)

### Example 1: Coin

<!-- TODO link to access control for mint -->
```move=
address 0x2 {
module M {
    // We do not want the Coin to be copied because that would be duplicating this "money",

>     // so we do not give the struct 'copy'

Reviewer: ```suggestion
    // so we do not give the struct the 'copy' ability.
```

>     // Similarly, we do not want programmers to destroy coins, so we do not give the struct 'drop'

Reviewer: ```suggestion
    // Similarly, we do not want programmers to destroy coins, so we do not give the struct the 'drop' ability.
```

    // However, we *want* users of the modules to be able to store this coin in persistent global

>     // storage, so we grant the struct the ability 'store'

Reviewer: ```suggestion
    // storage, so we grant the struct the ability 'store'.
```
